tiny_cnn/layers/arithmetic_layer.h

```diff
@@ -51,8 +51,8 @@ class elementwise_add_layer : public layer {
     }
 
     void forward_propagation(cnn_size_t worker_index,
-                             const std::vector<vec_t*>& in_data,
-                             std::vector<vec_t*>& out_data) override {
+                             const vec_t** in_data,
+                             vec_t** out_data) override {
```

Comment:
@patrikhuber Thank you for the lucid indications and I'm so sorry for delayed response.

> I am not really in favour of these changes. Do we really have to go into "2 star programming"? One of the reasons I like and use tiny-cnn is because of it's modern, clear code, and if we start introducing these pointers-to-pointers, everything becomes much harder to reason about.

Generally speaking, modifications done by this PR extensively spoiles conciseness and esthetics of the library, so I appreciate your concern.

But I still prefer vec_t** type over std::vector<vec_t*>& type because vector container allows resizing although it isn't needed and shouldn't be allowed in this case. And checking size of container isn't performed either so I think pointers of pointers is sufficient.

There are cons and pros with vector and pointer, so opinions vary. First of all, using pointers is dangerous.

> 1) Since we're passing around references, I don't see why there would be dynamic allocations in the first place? If there are, can we not solve it in a different way? The memory has to be allocated somewhere, can we not use ::reserve() there, for example? 

If I understand correctly, dinamic memory allocations are in layer::forward and layer::backward methods. STL vector container type local variables are used in there so small sized containers are frequently constructed/destructed. C++ style solution for this is using stack storage allocator but it's not part of the standard yet. Another solution is using thread_local keyword for those local vector variables. But simple pointer array is very simple and it should meet the demand but changing the prototypes makes a lot of diff...

> 
2) If this is the only way, can you show a speed difference of these isolated changes, showing that the loss of clarity and ability to reason about the code would be worth it?

I've attached the measurement.
[tiny-cnn-bench-beru-20160625.zip](https://github.com/nyanp/tiny-cnn/files/333043/tiny-cnn-bench-beru-20160625.zip)
I used below branch.
https://github.com/beru/tiny-cnn/commits/2star_loathed

Right now, I can only use poor spec laptop, so it's hard to get constant result and it takes hours to measure and the results may be not so accurate but I guess marginal speedup is possible with the change. Other people can also check the difference with the branch. I'll measure the difference with desktop PC at the end of this month.

> 3) If we were to adopt these changes, please at least let us add a a clear documentation to each function and the parameters.

Sure, and there're a lof of things I need to do such as separating AVX optimized code, merging #181 , write tests and descriptions... so I will improve this PR slowly.